### ISSUE TITLE
Handle missing indices in geometry+index constructors

### ISSUE THREAD

**Issue by caspervdw** [pygeos/pygeos#345]
We currently have mixed behaviour on the handling of missing indices in the geometry creation functions.

I'd like to decide on this and make it consistent. This already arised in #322.

Do we want this?

```
>>> pygeos.linestrings([[0.1, 0.1], [0.2, 0.2]], indices=[1, 1])
[None, Geometry("LINESTRING (0.1 0.1, 0.2 0.2)")]
```

Or this?

```
pygeos.multipoints([pygeos.Geometry("POINT (1 1)")] * 2, indices=[1, 1])
>>> array([<pygeos.Geometry MULTIPOINT EMPTY>,
                 <pygeos.Geometry MULTIPOINT (1 1, 1 1)>], dtype=object)
```

We basically have 3 options:
- raise an exception (like what is done when you supply 1 coordinate for a linestring)
- insert an EMPTY geometry (but; LINEARRING EMPTY does not exist, so what to do there?)
- insert a None

I am tending towards the first option. It is always better to make it more strict, then you can keep backwards compatibility when changing it later. What do you think @jorisvandenbossche @brendan-ward ?

**Comment by jorisvandenbossche** [comment]
I agree that for now being strict (raising an error) is fine, as we can always relax later on, while the other way around is more difficult. 

Long term, I think it would be nice to be able to create arrays this way that *do* contain empty and missing geometries, though (eg if you want to use those functionalities to roundtrip / serialize purposes). 
Now, both probably won't be possible, since we only have a missing index to signal this. When we have to choose one, I think it might be most logical to go for an empty geometry (so the roundtripping with the combo of those functions with `get_parts`/`get_coordinates` and `return_index=True` works for all geometries). If you want to support missing values in your roundtripping, you would then need to keep track of that separately (eg with a mask).

**Comment by jorisvandenbossche** [comment]
Relevant discussion from @brendan-ward at https://github.com/pygeos/pygeos/pull/322#issuecomment-805043013:

> I'm a bit torn on the handling of empty geometries / `None` / exceptions here.  Ideally, it would also be good for this to be consistent with #290, and either always return `None` when there are no indices for that output record or always return empty of that particular geometry type.  I do think that `None` seems more reasonable in this case, so I agree with the first part of the proposal (and apply that back to the geometry collections too).  An exception doesn't seem appropriate in this case.

**Comment by jorisvandenbossche** [comment]
> (but; LINEARRING EMPTY does not exist, so what to do there?)

I think in that case raising an error would still be fine. I know that's a bit inconsistent, but on the other hand, you would also never get missing indices for linearrings from using pygeos functionality like `get_rings`+`get_coordinates` (assuming the user starts with polygons, and is not actually working with linearrings directly).

**Comment by caspervdw** [comment]
It seems that there is something to say for EMPTY and for None. In that case I think we should make it a keyword argument `on_missing_index={"none","empty","raise"}`.

If we decide on filling in None or Empty, then we should *also* decide what to do with None/Empty geometries at the end. e.g. if you do `get_coordinates([Geometry("LINESTRING (0.1 0.1, 0.2 0.2)"), None], return_index=True)` the information that it ended with `None` is lost. Options to solve that:

- supply a minimum size (like in https://numpy.org/doc/stable/reference/generated/numpy.bincount.html)
- add the option of supplying an `out` array

I'd like to opt for the `out` array, as it is very similar to the ufunc API. If `out` is supplied, we have the additional option of doing nothing (leaving the `out` value whatever it was), next to replacing it with None or EMPTY. So that would amount to `on_missing_index={"ignore","none","empty","raise"}`.

To be clear, I'll go for the exception option in the coming release. This will take significant work to get right.

**Comment by caspervdw** [comment]
After some discussion in #346 I think we could also do this:

- if an index is missing, insert None (or do nothing if an “out” array is specifically given)
- if an index is present, but refers to a missing geometry, always construct a geometry (in case of all-None geometry inputs: yield EMPTY)

We could extend this principle to simple geometry construction as well, so that we can generate empty geometries there too.

- if an input coordinate is NaN, ignore it

Avoiding NaN coordinates is a good thing: GEOS allows it, but functions can not handle them.

**Comment by caspervdw** [comment]
Quoting from @brendan-ward (#346):

> Provided I'm following #345 correctly, it seems like the handling of None should be consistent with and without indices. I'm currently seeing this:
> 
> >>> pygeos.polygons([None])
> array([None], dtype=object)
> >>> pygeos.polygons([None], indices=[0])
> array([<pygeos.Geometry POLYGON EMPTY>], dtype=object)
> 
> Handling that would be an API change for the non-indices version of polygons, but it seems challenging to communicate that this behavior is toggled by the use of indices.

We did not yet reach a conclusion here. My latest proposal can be seen in the previous post. I'd welcome your input!

You are pointing out an inconsistency between the "normal" `polygons` and the "indices" `polygons`. I think the inconsistency should be solved on the "normal" `polygons` side. Namely, it is also inconsistent with the collection constructors, like `geometrycollections`. For example:

```
In [10]: pygeos.polygons([None])
Out[10]: array([None], dtype=object)
In [11]: pygeos.geometrycollections([None])
Out[11]: <pygeos.Geometry GEOMETRYCOLLECTION EMPTY>
```

The problem here is that the "normal" `polygons` takes shells + holes, while the "indices" one just takes `linearrings`. 
I think we should deprecate the `holes` altogether. I'll propose something in #346.

**Comment by caspervdw** [comment]
I'd like to round up this discussion by proposing the following:

**Add an `out` parameter**
Just as with the ufuncs, it is sometimes efficient to use an already-allocated array. The geometry constructors could use these.
If an index is missing, nothing will happen in the array. If an index is present, but referring to a `None` geometry, an empty geometry will be constructed and put into the out array.

**Handle missing indices in case no empty array is provided**
Raise an exception (like now)

@jorisvandenbossche @brendan-ward
